### PR TITLE
Vectors of pointers

### PR DESCRIPTION
**Pull request by calebzulawski**
* Adds vectors of pointers
* Changes scatter/gather implementation (but doesn't introduce new pointer API, yet)
* Modifies `Simd::cast`, since all casts are no longer valid (e.g. casting `f32` to a pointer)

### CODE REVIEW

File: crates/core_simd/src/cast.rs
Comment (programmerjake) [inline crates/core_simd/src/cast.rs:54]: I would consider that questionably safe, pointer to integer transmutes are almost certainly UB:
https://github.com/rust-lang/unsafe-code-guidelines/issues/286

Comment (programmerjake) [reply crates/core_simd/src/cast.rs:54]: also, the safety comment is the wrong way around, you're transmuting pointers to integers here, which is probably UB.

File: crates/core_simd/src/elements/const_ptr.rs
Comment (programmerjake) [inline crates/core_simd/src/elements/const_ptr.rs:64]: this really should use either a compiler intrinsic for vector wrapping add, or use `Simd::cast`, once casting is fixed.

Comment (programmerjake) [reply crates/core_simd/src/elements/const_ptr.rs:64]: also, if `addr` is supposed to be like `<*const T>::addr`, that specifically returns an integer that can't be converted back to a pointer. so that side also needs to be `cast`

File: crates/core_simd/src/cast.rs
Comment (thomcc) [reply crates/core_simd/src/cast.rs:54]: It's no longer almost certainly UB, but it does lose provenance (and `as`/casting should not): https://doc.rust-lang.org/nightly/src/core/ptr/const_ptr.rs.html#183-185

File: crates/core_simd/src/elements/const_ptr.rs
Comment (programmerjake) [inline crates/core_simd/src/elements/const_ptr.rs:24]: the docs need some elaboration -- especially specifically mentioning that the result can't be cast back to a pointer and used.
https://github.com/rust-lang/rust/blob/10f4ce324baf7cfb7ce2b2096662b82b79204944/library/core/src/ptr/const_ptr.rs#L178

File: crates/core_simd/src/cast.rs
Comment (programmerjake) [reply crates/core_simd/src/cast.rs:54]: i meant its UB to use the result by converting it back to a pointer and expecting it to be valid.

Comment (thomcc) [reply crates/core_simd/src/cast.rs:54]: I see. That is still UB (for a while, the transmute itself was probably UB too).

Comment (calebzulawski) [reply crates/core_simd/src/cast.rs:54]: My understanding is that once you convert a pointer to an integer, it loses provenance no matter what, and you need to use `with_addr` to inherit an existing pointer's provenance.  So this particular line (converting a pointer to an integer) should be fine, I think.

File: crates/core_simd/src/elements/const_ptr.rs
Comment (calebzulawski) [reply crates/core_simd/src/elements/const_ptr.rs:64]: This is just a rewrite of what we already had.  Definitely needs to use an intrinsic to be correct.

File: crates/core_simd/src/elements/const_ptr.rs
Comment (calebzulawski) [reply crates/core_simd/src/elements/const_ptr.rs:24]: Thought it would be better to reference the underlying function it's equivalent to, especially considering it's likely to change sometime before stabilization

File: crates/core_simd/src/cast.rs
Comment (programmerjake) [reply crates/core_simd/src/cast.rs:54]: that's actually incorrect -- on most platforms using `as` to cast a pointer to an integer exposes the pointer such that `as` casts from integers back to pointers using that address produces the original pointer including provenance. transmute doesn't do that.

because these functions are named `cast` and intended to function like `as`, they need to have the same pointer exposing behavior as `as`. you don't get that with `transmute`

File: crates/core_simd/src/elements/const_ptr.rs
Comment (programmerjake) [reply crates/core_simd/src/elements/const_ptr.rs:24]: yes, but imho we need to point out that you can't use it like `ptr as usize` which allows getting `ptr` back by `as` casting the resulting integer back to a pointer. otherwise people who only read the docs of this function may assume it's `ptr as usize`, leading them to write code with UB.

File: crates/core_simd/tests/pointers.rs
Comment (programmerjake) [inline crates/core_simd/tests/pointers.rs:26]: testing `wrapping_offset` on a `*mut ()` is kinda pointless since it always returns the input pointer, since `()` is a ZST. imho it would be better to test on `*mut u32` since it has both a nonzero pointee size (avoiding the output always being the same as the input), and a non-one pointee size (catching the potential bug where rustc forgot to multiply the offset by the pointee size)

File: crates/core_simd/src/cast.rs
Comment (calebzulawski) [reply crates/core_simd/src/cast.rs:54]: I realized my confusing when reviewing the docs for `addr`:
> This is equivalent to `self as usize`, which semantically discards provenance and address-space information. To properly restore that information, use [with_addr](https://doc.rust-lang.org/std/primitive.pointer.html#method.with_addr) or [map_addr](https://doc.rust-lang.org/std/primitive.pointer.html#method.map_addr).

The bit about `self as usize` is incorrect, then?

Comment (programmerjake) [reply crates/core_simd/src/cast.rs]: > I realized my confusing when reviewing the docs for `addr`:
> 
> > This is equivalent to `self as usize`, which semantically discards provenance and address-space information. To properly restore that information, use [with_addr](https://doc.rust-lang.org/std/primitive.pointer.html#method.with_addr) or [map_addr](https://doc.rust-lang.org/std/primitive.pointer.html#method.map_addr).
> 
> The bit about `self as usize` is incorrect, then?

yes, when referring to current `ptr as usize` semantics...that said, iirc they want to move to `as` being equivalent to `addr` rather than `expose_addr` at some point (requires edition boundary).

File: crates/core_simd/src/elements/const_ptr.rs
Comment (programmerjake) [inline crates/core_simd/src/elements/const_ptr.rs:89]: actually, `<*const T>::addr` uses `transmute` right now, it'll eventually be its own intrinsic, not a cast.

Comment (calebzulawski) [reply crates/core_simd/src/elements/const_ptr.rs:89]: Interesting, on stable it's `as`, but `transmute` now on nightly.

Comment (programmerjake) [reply crates/core_simd/src/elements/const_ptr.rs]: iirc that's because `transmute` was recently changed so ptr -> int isn't UB.

**Review by programmerjake — COMMENTED** [review]
lgtm, assuming this is changed once simd int <-> ptr casts land in rustc

**Comment by calebzulawski** [comment]
New intrinsics merged, so this is ready to go.

**Comment by workingjubilee** [comment]
Huzzah.

File: crates/core_simd/src/elements/const_ptr.rs
Comment (programmerjake) [inline crates/core_simd/src/elements/const_ptr.rs:64]:
```suggestion
    /// Equivalent to calling [`pointer::wrapping_sub`] on each lane.
```

File: crates/core_simd/src/elements/mut_ptr.rs
Comment (programmerjake) [inline crates/core_simd/src/elements/mut_ptr.rs:61]:
```suggestion
    /// Equivalent to calling [`pointer::wrapping_sub`] on each lane.
```

**Review by programmerjake — APPROVED** [review]
once those two `wrapping_sub` doc comments are fixed, lgtm

**Review by programmerjake — APPROVED** [review]

**Comment by workingjubilee** [comment]
Apologies, going to try to get to this this week.

**Comment by calebzulawski** [comment]
Okay thanks!

File: crates/core_simd/src/cast.rs
Comment (workingjubilee) [inline crates/core_simd/src/cast.rs:40]:
```suggestion
/// Supporting trait for `Simd::cast_ptr`.  Typically doesn't need to be used directly.
///
/// # Safety
/// Implementing this trait asserts that the type is a valid vector element for the `simd_cast_ptr`
/// intrinsic.
pub unsafe trait SimdCastPtr: SimdElement {}
```

Is there a reason we can't just use the bounds that std uses to guarantee the Pointee has the "Metadata" of `()` (and is thus a thin pointer)?

File: crates/core_simd/src/cast.rs
Comment (workingjubilee) [inline crates/core_simd/src/cast.rs:45]: This bound doesn't seem correct since I think this allows fat pointers.

File: crates/core_simd/src/eq.rs
Comment (workingjubilee) [inline crates/core_simd/src/eq.rs:77]: This seems okay. Are there any caveats, e.g. if the pointers get evaluated at const time, that we should document?

File: crates/core_simd/src/ord.rs
Comment (workingjubilee) [inline crates/core_simd/src/ord.rs:312]: The notion of clamping pointers seems somewhat dubious...?

File: crates/core_simd/src/vector.rs
Comment (workingjubilee) [inline crates/core_simd/src/vector.rs:258]: Huh, so this happens because `SimdElement` is now either `SimdCast` xor `SimdPtrCast`, right?

File: crates/test_helpers/src/lib.rs
Comment (workingjubilee) [inline crates/test_helpers/src/lib.rs:174]: This does look a bit nicer.

File: crates/test_helpers/src/lib.rs
Comment (workingjubilee) [inline crates/test_helpers/src/lib.rs:237]: Hm, maybe it would be nice if we could just `try_collect` here instead...? Is that hypothetically possible?

File: crates/test_helpers/src/lib.rs
Comment (workingjubilee) [inline crates/test_helpers/src/lib.rs:201]: huzzah less bounds.

**Review by workingjubilee — COMMENTED** [review]
Thank you for this! Sorry for being so distracted. Mostly things look good but I have a few questions?

File: crates/core_simd/src/vector.rs
Comment (calebzulawski) [reply crates/core_simd/src/vector.rs:258]: Yeah.  But really, it has nothing to do with `SimdPtrCast` (if rust ever added a numeric pointer type for some reason, it could implement both).  This could be thought of as `SimdNumeric`, but specific to casting.

File: crates/core_simd/src/cast.rs
Comment (calebzulawski) [reply crates/core_simd/src/cast.rs]: I did change `SimdElement` to only implement on thin pointers, for the time being.

File: crates/core_simd/src/cast.rs
Comment (calebzulawski) [reply crates/core_simd/src/cast.rs]: I changed it to require `T: Pointee<Metadata = ()>`

File: crates/core_simd/src/vector.rs
Comment (workingjubilee) [reply crates/core_simd/src/vector.rs:259]: Ahhh okay.

File: crates/core_simd/src/cast.rs
Comment (workingjubilee) [reply crates/core_simd/src/cast.rs]: I think that combined, those two changes are correct, and I agree that I don't see how the metadata casting would be valid otherwise if the metadata did not match, except for the special case of a cast that strips metadata (i.e. takes any `U` to `()`) (at which point we don't need that to be a "SIMD cast", really) (it's fine to not model every possible action, especially if the action is dubious).

File: crates/core_simd/src/ord.rs
Comment (calebzulawski) [reply crates/core_simd/src/ord.rs:312]: It is odd, but pointers already implement `Ord` 🤷‍♂️

File: crates/core_simd/src/eq.rs
Comment (calebzulawski) [reply crates/core_simd/src/eq.rs:77]: I'm not sure it has any caveats beyond what's already true for `const fn` or whatever.  Did you have anything specific in mind?

File: crates/core_simd/src/vector.rs
Comment (calebzulawski) [reply crates/core_simd/src/vector.rs:259]: I think I will change it to `T: SimdCast<U>`, to match the updated `SimdPtrCast` (and in case we want to use similar bounds in the future).

Comment (calebzulawski) [reply crates/core_simd/src/vector.rs:259]: Actually, I realized this would require a quadratic number of implementations... It's probably fine how it is.

File: crates/test_helpers/src/lib.rs
Comment (calebzulawski) [reply crates/test_helpers/src/lib.rs:237]: I think so, I didn't know about `try_collect` until now.  I'm not sure it's worth bringing in an unstable feature just for that though.

Comment (workingjubilee) [reply crates/test_helpers/src/lib.rs:237]: Yeah, that's fair. I am mostly curious if it's possible for users of `try_collect` to experiment with.

Comment (calebzulawski) [reply crates/test_helpers/src/lib.rs:237]: I just realized, I think this is the opposite of `try_collect`.  The iterator items can't error, the conversion of the collection is fallable (`Vec` to array conversion)

Comment (workingjubilee) [reply crates/test_helpers/src/lib.rs:237]: Oh okay! Whoops.

File: crates/core_simd/src/eq.rs
Comment (workingjubilee) [reply crates/core_simd/src/eq.rs:77]: I am thinking of the pointer indeterminacy problem during CTFE. Things that would share addresses during runtime may not during compile-time function execution, and vice versa.

Comment (calebzulawski) [reply crates/core_simd/src/eq.rs:77]: Honestly, not sure what the implications would be other than since the SIMD intrinsics aren't const (yet) so we know any pointers for now will be runtime only.  If you do something bad with CTFE before putting it in a vector, that's probably out of scope for now.

Comment (calebzulawski) [reply crates/core_simd/src/eq.rs:77]: I think it's fair to assume all vector operations should mirror scalar operations in that regard, whatever the implications are.

Comment (workingjubilee) [reply crates/core_simd/src/eq.rs:77]: Fair enough. I think it's mostly something we should keep in mind while exploring the API surface.

File: crates/core_simd/src/intrinsics.rs
Comment (RalfJung) [inline crates/core_simd/src/intrinsics.rs:163]: It look like currently no tests actually call any of these intrinsics, or am I missing something?

That would explain why portable-simd tests still pass in Miri where these intrinsics are not implemented.

Comment (calebzulawski) [reply crates/core_simd/src/intrinsics.rs:163]: Ah, looking at the tests, I did omit some of the very simple ones like this.  They shouldn't be hard to add.

**Pull request merged**
